### PR TITLE
Provide example data

### PR DESCRIPTION
**Pull request by Teschl**
In this pull request, I added the following functions:

- `topotoolbox.load_dem()` enter DEM name as argument, it will download the tif if it isn't already cached.
- `topotoolbox.get_dem_names()` will return a list of all available tifs for `load_dem()`. if we add a .txt which contains all names to the repository which contains the tifs, we can implement it more dynamically. (right now it's just a list in the .py file)
- `topotoolbox.read_tif()` is a wrapper for creating a GridObject. Could be a cleaner way to create a GridObject inspired by pandas (pandas.read_csv creates a pandas.Dataframe).

Since the function `get_save_location()` in `utils.py` is not listed in `__all__` it's not automatically imported when using the package. This way it's "private" and not usable for the user by default.

I also fixed some line length issues in the mixin files for the GridObject. The maximum line length is 79 chars.

### CODE REVIEW

**Comment by Teschl** [comment]
Regarding the repository that contains the examples, right now, it's the [https://github.com/wschwanghart/DEMs](https://github.com/wschwanghart/DEMs) repository. So we might want to fork it to `https://github.com/TopoToolbox/DEMs`. Then we also should add a file like `contents.txt` in which we store all names of the provided DEMs. That means we could remove the hardcoded `DEM_NAMES` of `utils.py` and generate them dynamically.

If implemented, this should resolve #9 .

**Comment by wkearn** [comment]
https://github.com/TopoToolbox/DEMs

File: src/topotoolbox/utils.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    return dem
+
+
+def get_save_location() -> str:
+    """Generates filepath to file saved in cache.
+
+    Args:
+        data_home (str, optional): name of directory in cache. Defaults to None
+        which results in "topotoolbox" as name.
```
Comment (wkearn) [inline src/topotoolbox/utils.py:76]: This argument doesn't exist.

File: src/topotoolbox/utils.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    url = f"{DEM_SOURCE}/{dem}.tif"
+
+    if cache:
+        cache_path = os.path.join(get_save_location(), f"{dem}.tif")
+
+        if not os.path.exists(cache_path):
+            urlretrieve(url, cache_path)
+
+        full_path = cache_path
```
Comment (wkearn) [inline src/topotoolbox/utils.py:62]: Can we provide a `clear_cache` function or something similar to allow users to empty this directory if they decide they don't need the data anymore?

File: src/topotoolbox/utils.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+def read_tif(path: str) -> GridObject:
+    """Generate a new GridObject from a .tif file.
+
+    Args:
+        path (str): path to .tif file.
+
+    Returns:
+        GridObject: A new GridObject of the .tif file.
+    """
+    return GridObject(path)
```
Comment (wkearn) [inline src/topotoolbox/utils.py:25]: Would it make sense to read the data through rasterio in here and then use `read_tif` in the `GridObject` constructor? That might let us use the `GridObject` constructor to dispatch to different methods for loading data depending on the arguments provided to it. That might be a problem for another pull request though.

**Review by wkearn — CHANGES_REQUESTED** [review]
Looks pretty good, @Teschl. This should be sufficient for our testing and documentation purposes at the moment. I suggested a few changes that make sense right now. 

In the future, we might want to think about other ways that users might want to access data and how we can accomodate them. The MATLAB TopoToolbox [provides](https://raw.githubusercontent.com/TopoToolbox/topotoolbox/2ef445126fd0e6c9c336dce19a205b2a2e62f8a2/IOtools/readopentopo.m), for example, a way to access the [OpenTopography API](https://opentopography.org/developers) that could be nice to have across our packages.

**Comment by wschwanghart** [comment]
Excellent. I think it's generally a good idea to host also the example DEMs
on git although git is not really made for data hosting. However, so far it
supports file sizes up to 50 Mb. We might also consider this here:
https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage

Am Di., 28. Mai 2024 um 19:57 Uhr schrieb William Kearney <
***@***.***>:

File: src/topotoolbox/utils.py (hunk 3 above)
Comment (Teschl) [reply src/topotoolbox/utils.py:25]: Yes, I think that would be better. I think moving the `load_dem` to `grid_object.py` would be a good move. Also, moving `gen_random()`, [...] out of the class will improve usability. (calling `topotoolbox.gen_random()` is better than calling `topotoolbox.GridObject.gen_random())

Therefore, calling the `__init__()` of the GridObject should require passing all attributes and having them prepared beforehand.

These are quite a few changes, though, so that should definitely be done in a different pull request.

Also, just to add to that: while implementing this pull request, I was looking at a few other repositories like `pandas` and `seaborn`. Nobody seems to be using mixins. It seems to be convention to have very long files (2000 lines is not uncommon like [here](https://github.com/pandas-dev/pandas/blob/main/pandas/core/arrays/categorical.py)). So after all, maybe we should get rid of them.
The only issue I see with that in our case is, that we are importing some files that only exist when building (`from ..grid import grid_identifyflats`). So my IDE is always yelling at me that some files don't exist. That's circumnavigated by using mixins.

File: src/topotoolbox/utils.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    Args:
+        dem (str): Name of dem about to be downloaded
+        cache (bool, optional): If true the dem will be cached. 
+        Defaults to True.
+
+    Returns:
+        GridObject: A GridObject generated from the downloaded dem.
+    """
+
+    url = f"{DEM_SOURCE}/{dem}.tif"
```
[the remark this answers is not in the document]
Comment (Teschl) [reply src/topotoolbox/utils.py:54]: I was thinking we could also add a function like `store_dem()` that caches DEMs as .tif files. That will require the opposite of `read_tif()`: `write_tif()`.
Then calling `load_dem()` will either get the file with the respective name from the cache or download it if it's available and not in the cache already.

**Comment by Teschl** [comment]
I have addressed the more bug-fix related requested changes. 

> Would it make sense to read the data through rasterio in here and then use read_tif in the GridObject constructor? That might let us use the GridObject constructor to dispatch to different methods for loading data depending on the arguments provided to it. That might be a problem for another pull request though.

and maybe adding `store_dem()` can be handled in a different pull request.
